Tutorial - Bubble A/B tests (#7500)
PostHog/posthog.com: contents/tutorials/bubble-ab-tests.md

---
title: "How to run A/B tests in Bubble with PostHog"
date: 2024-01-11
author: ["lior-neu-ner"]
showTitle: true
sidebar: Docs
tags: ['experimentation']
---

import { ProductScreenshot } from 'components/ProductScreenshot'
import EventsInPostHogLight from '../images/tutorials/bubble-ab-tests/events-in-posthog-light.png'
import EventsInPostHogDark from '../images/tutorials/bubble-ab-tests/events-in-posthog-dark.png'
import TestSetupLight from '../images/tutorials/bubble-ab-tests/test-setup-light.png'
import TestSetupDark from '../images/tutorials/bubble-ab-tests/test-setup-dark.png'

> [Bubble](https://bubble.io/) is a great tool for building marketing websites. However, sometimes you may be unsure if a change will actually improve your conversion rate. This is where [A/B testing](/ab-testing) is helpful. It enables you to test and compare the results of your changes.

Reviewer: ```suggestion
[Bubble](https://bubble.io/) is a great tool for building marketing websites. However, sometimes you may be unsure if a change actually improves your conversion rate. This is where [A/B testing](/ab-testing) is helpful. It enables you to test and compare the results of your changes.
```

    posthog.init('<ph_project_api_key>',{api_host:'<ph_instance_address>'})
</script>
```

With the snippet copied, go to your Bubble site settings by clicking on the icon in the left-hand menu. If you haven’t already, sign up for at least the **Starter** site plan. This enables you to add custom code.

Go to the **SEO / metatags** tab in site settings. Paste your PostHog snippet in the **Script/meta tags in header** section. Then, deploy your site to live.

![How to add PostHog to Bubble](../images/tutorials/bubble-ab-tests/adding-posthog.mp4)

## 2. Capture a custom event

> We're going to create an A/B test comparing how the text of a button affects it's click-through rate. To compare the results, we first capture a custom event and then set it as our [goal metric](/product-engineers/ab-testing-guide-for-engineers#1-a-clear-measurable-goal) for our A/B test. 

Reviewer: ```suggestion
We're going to create an A/B test comparing how a button's text affects its click-through rate. To compare the results, we first capture a custom event and then set it as our [goal metric](/product-engineers/ab-testing-guide-for-engineers#1-a-clear-measurable-goal) for our A/B test. 
```

> Here are the steps to follow to capture this event:

Reviewer: ```suggestion
To capture this event:
```

1. Install the [Bubble Toolbox plugin](https://bubble.io/plugin/toolbox-1488796042609x768734193128308700). This enables us to run custom JavaScript code when the button is pressed.

2. Add a new button to your Bubble site. Click on the button and then on **Add workflow** in the popup menu.

3. Click on **add an action**. In the menu that appears, click on **Plugins** and then **Run javascript**. This will open a new menu where you can add JavaScript code.

4. Add the following code under the **Script** heading:

```js
window.posthog.capture("ab_test_button_clicked")
```

![Capture custom event in Bubble](../images/tutorials/bubble-ab-tests/capture-custom-event.mp4)

Finally, deploy your changes to live and then click your new button to [see the event in PostHog](https://us.posthog.com/events).

<ProductScreenshot
  imageLight={EventsInPostHogLight} 
  imageDark={EventsInPostHogDark} 
  alt="Events captured in PostHog" 
  classes="rounded"
/>

## 3. Create an A/B test in PostHog

> If you haven't done so already, you'll need to [upgrade](https://us.posthog.com/organization/billing) your PostHog account to include A/B testing. This requires entering your credit card details, but don't worry, we have a [generous free tier](/pricing) of 1 million free requests – so you won't be charged anything yet.

Reviewer: You already say it is free 
```suggestion
If you haven't done so already, you'll need to [upgrade](https://us.posthog.com/organization/billing) your PostHog account to include A/B testing. This requires entering your credit card details, but don't worry, we have a [generous free tier](/pricing) of 1 million requests per month – so you won't be charged anything yet.
```

Next, go to the [A/B testing tab](https://us.posthog.com/experiments) and create an A/B test by clicking the **New experiment** button. Add the following details to your experiment:

1. Name it "Bubble button experiment".
2. Set "Feature flag key" to `bubble-button-experiment`.

> 3. Under the experiment goal, select the `ab_test_button_clicked` we created in the previous step.

Reviewer: ```suggestion
3. Under the experiment goal, select the `ab_test_button_clicked` event we created in the previous step.
```

4. Use the default values for all other fields.

Click "Save as draft" and then click "Launch".

<ProductScreenshot
  imageLight={TestSetupLight} 
  imageDark={TestSetupDark} 
  alt="Experiment setup in PostHog" 
  classes="rounded"
/>

## 4. Implement the A/B test in Bubble

> The final step is to add the experiment code in Bubble. Here we'll add code that does the following:

Reviewer: It is slightly unclear where "here" is.
```suggestion
The final step is to add the experiment code in Bubble. We need to add code that does the following:
```

To do this, we add another workflow event. Go to the [Workflow tab](https://manual.bubble.io/help-guides/getting-started/navigating-the-bubble-editor/tabs-and-sections/workflow-tab) in Bubble. Create a new **Page is loaded** event with an action to run custom JavaScript code (like we did for [capturing our custom event](#2-capture-a-custom-event)). 

Use the following code to change the title of the button based on the value of the feature flag:

```js
posthog.onFeatureFlags(() => {
   // Find the button you'd like to change. In our case, we change the text of the "Sign Up" button
   var button = Array.from(document.querySelectorAll('button')).find(el => el.textContent === 'Sign Up');

   if (button) {
      const flagValue = posthog.getFeatureFlag('bubble-button-experiment')
      if (flagValue === 'control') {
        button.textContent = "This is the title for CONTROL group"
      } else if (flagValue === 'test') {
        button.textContent = "This is the title for TEST group"
      }
   }
});    
```

Finally, deploy your changes to live.

![Add script in Framer](../images/tutorials/bubble-ab-tests/add-flag-code.mp4)

> That's it! Your A/B test is now live. Half your users will see the updated button text and PostHog will track whether it has an impact on your click-through rate. You can [view your test results](/docs/experiments/testing-and-launching#viewing-experiment-results) on the experiment page in PostHog.

Reviewer: ```suggestion
That's it! Your A/B test is now live. PostHog will split your users so half see the updated button text and tracks whether it has an impact on your click-through rate. You can [view your test results](/docs/experiments/testing-and-launching#viewing-experiment-results) on the experiment page in PostHog.
```
